Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,1486 @@
+      toFinsupp_smul with
+    toDistribMulAction := distribMulAction }
+
+instance faithfulSMul {S} [AddCommMonoid k] [SMulZeroClass S k] [FaithfulSMul S k] [Nonempty G] :
+    FaithfulSMul S (SkewMonoidAlgebra k G) where
+  eq_of_smul_eq_smul {_s₁ _s₂} h :=
+    eq_of_smul_eq_smul fun a : G →₀ k => congr_arg toFinsupp (h ⟨a⟩)
+
+instance isScalarTower {S₁ S₂} [AddCommMonoid k] [SMul S₁ S₂] [SMulZeroClass S₁ k]
+    [SMulZeroClass S₂ k] [IsScalarTower S₁ S₂ k] : IsScalarTower S₁ S₂ (SkewMonoidAlgebra k G) :=
+  ⟨by
+    rintro _ _ ⟨⟩
+    simp_rw [← ofFinsupp_smul, smul_assoc]⟩
+
+instance smulCommClass {S₁ S₂} [AddCommMonoid k] [SMulZeroClass S₁ k] [SMulZeroClass S₂ k]
+    [SMulCommClass S₁ S₂ k] : SMulCommClass S₁ S₂ (SkewMonoidAlgebra k G) :=
+  ⟨by
+    rintro m n ⟨f⟩
+    simp_rw [← ofFinsupp_smul, smul_comm m n f]⟩
+
+instance isCentralScalar {S} [AddCommMonoid k] [SMulZeroClass S k] [SMulZeroClass Sᵐᵒᵖ k]
+    [IsCentralScalar S k] : IsCentralScalar S (SkewMonoidAlgebra k G)  :=
+  ⟨by
+    rintro _ ⟨⟩
+    simp_rw [← ofFinsupp_smul, op_smul_eq_smul]⟩
+
+variable {M α : Type*} [Monoid G] [AddCommMonoid M] [MulAction G α]
+
+/-- Scalar multiplication acting on the domain.
+
+This is not an instance as it would conflict with the action on the range.
+See the `instance_diamonds` test for examples of such conflicts. -/
+def comapSMul [AddCommMonoid M] : SMul G (SkewMonoidAlgebra M α) where smul g := mapDomain (g • ·)
+
+attribute [local instance] comapSMul
+
+theorem comapSMul_def (g : G) (f : SkewMonoidAlgebra M α) : g • f = mapDomain (g • ·) f := rfl
+
+@[simp]
+theorem comapSMul_single (g : G) (a : α) (b : M) : g • single a b = single (g • a) b :=
+  mapDomain_single
+
+/-- `comapSMul` is multiplicative -/
+def comapMulAction : MulAction G (SkewMonoidAlgebra M α) where
+  one_smul f := by rw [comapSMul_def, one_smul_eq_id, mapDomain_id]
+  mul_smul g g' f := by
+    rw [comapSMul_def, comapSMul_def, comapSMul_def, ← comp_smul_left, mapDomain_comp]
+
+attribute [local instance] comapMulAction
+/-- This is not an instance as it conflicts with `SkewMonoidAlgebra.distribMulAction`
+  when `G = kˣ`. -/
+def comapDistribMulActionSelf [AddCommMonoid k] :
+    DistribMulAction G (SkewMonoidAlgebra k G) where
+  smul_zero g := by
+    ext
+    simp only [comapSMul_def, smul_eq_mul, sum_zero_index, toFinsupp_zero, Finsupp.coe_zero,
+      Pi.zero_apply]
+  smul_add g f f' := by
+    ext
+    simp only [comapSMul_def, mapDomain_add]
+
+
+end DerivedInstances
+
+section MiscTheorems
+
+variable [Semiring k]
+
+theorem mul_apply [DecidableEq G] [Monoid G] [MulSemiringAction G k] (f g : SkewMonoidAlgebra k G)
+    (x : G) : (f * g) x = f.sum fun a₁ b₁ => g.sum fun a₂ b₂ =>
+      if a₁ * a₂ = x then b₁ * a₁ • b₂ else 0 := by
+  rw [mul_def, sum_apply]; congr; ext
+  rw [sum_apply]; congr; ext
+  apply single_apply
+
+open BigOperators
+
+theorem mul_apply_antidiagonal [Monoid G] (f g : SkewMonoidAlgebra k G) [MulSemiringAction G k]
```

Comment:
I am puzzled by the applicability of this lemma. It requires the monoid to be very particular (G = Z doesn't work), but the statement holds for `s : Set (G × G)` using the fact that `f` and `g`have finite support.

Reply:
This is a good point.

I can prove the following:
```
theorem mul_apply_antidiagonal' [Monoid G] (f g : SkewMonoidAlgebra k G) [MulSemiringAction G k]
    (x : G) (s : Finset (G × G)) (hs : ∀ {p : G × G}, p ∈ s ↔
    (p.1 ∈ f.support ∧ p.2 ∈ g.support ∧ p.1 * p.2 = x)) :
    (f * g) x = ∑ p in s, f p.1 * p.1 • g p.2
```
I think this fixes the issue you raised and should work when we consider the additive monoid `Z`. I still think this kind of statement is a bit weird, there might be a better way to phrase this type of theorem.

Reply:
Assuming the correction I proposed fixes your issue, do you think we should keep both versions? Or just the new one?